personal_assistant/commands.py

```diff
@@ -0,0 +1,270 @@
+import random
+from functools import wraps
+
+from personal_assistant.src.address_book.record import Record
+from personal_assistant.src.exceptions.exceptions import WrongInfoException, NoValue, WrongDate, WrongEmail, \
+    WrongAddress
+from personal_assistant.src.personal_assistant import personal_assistant
+from personal_assistant.src.utils import check_args, wrong_input_handling, get_contact, pre_check_addr
+
+
+def greeting():
+    ## select and play a greeting reaction
+    prompt = random.choice(["Hello!", "Hi!", "Greetings!"])
+    return prompt
+
+
+def cache_data(command):
+    @wraps(command)
+    def wrapper(*args, **kwargs):
+        result = command(*args, **kwargs)
+        personal_assistant.cache_data()
+        return result
+
+    return wrapper
+
+
+@cache_data
+@wrong_input_handling
+def add_contact(args):
+    check_args(args, WrongInfoException())
+
+    name = args[0]
+    phone = args[-1]
+
+    if name in list(personal_assistant.address_book.data.keys()):
+
+        book_entry = personal_assistant.address_book.data[name]
+        try:
+            if book_entry.find_phone(phone):
+                return "This phone number is already associated with this contact."
+        except:
+            pass
+
+        confirm = input("A contact with this name found. Update it? yes / no: ")
+
+        if confirm.lower() in ["yes", "1", "affirmative", "y"]:
+            book_entry.add_phone(args[-1])
+        else:
+            return "Cancelling contact addition."
+
+    else:
+        contact = Record(name)
+        contact.add_phone(args[-1])
+        personal_assistant.address_book.add_record(contact)
+
+    return "Contact added."
+
+
+@cache_data
+@wrong_input_handling
+def change_contact(args):
+    check_args(args, ValueError())
+
+    contact = get_contact(personal_assistant.address_book, args[0])
+    contact.edit_phone(args[1], args[-1])
+    return "Contact updated."
+
+
+@wrong_input_handling
+def show_phone(args):
+    check_args(args, NoValue())
+    contact = get_contact(personal_assistant.address_book, args[0])
+    found_phones = contact.list_str_rep(contact.phones)
+    found_phones = "; ".join(found_phones)
+    return f"{args[0]}'s phone numbers: {found_phones}"
+
+
+@wrong_input_handling
+def show_all():
+    add_phone_message = 'Enter "add <name> <number>" to add a contact.'
+    if not personal_assistant.address_book.data:
+        return "No contacts found. " + add_phone_message
+
+    all_records = []
+
+    for i, record in enumerate(personal_assistant.address_book.data.values()):
+        name = record.name.value
+        phones = '\n'.join(p.value for p in record.phones) if record.phones else ''
```

Comment:
it seems multiple phones breaks the pattern that is printed out :
![image](https://github.com/nazarsmith/ppap7856-pythoneers-brigade-prjct/assets/61598939/b611a746-2b16-4f29-a482-77841c90a3ad)
I haven't tested multiple emails, but looks like it'll be the same thing

Reply:
we can merge as is so that this refactor is available to everyone and fix up in the next version